### PR TITLE
Support pass by ref args in for direct actor calls

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

This adds support for pass by ref args in direct actor calls. Most of the complication is in implementing the optimization that allows calls to be reordered according to when dependencies are available (to avoid blocking on ray.get() of arguments).

The strategy is to send a Wait() IPC to the raylet for each object dependency set via the raylet client. The raylet will send back a notification via gRPC when the wait completes.

ray microbenchmark (on laptop):
```
single client get calls per second 19854.25 +- 6166.4
single client put calls per second 6764.97 +- 23.33
single client put gigabytes per second 12.39 +- 0.57
multi client put calls per second 10314.35 +- 864.0
multi client put gigabytes per second 6.59 +- 0.49
single client tasks sync per second 1901.01 +- 181.56
single client tasks async per second 4165.44 +- 325.17
multi client tasks async per second 4240.08 +- 28.5
single client actor calls sync per second 1852.27 +- 91.86
single client actor calls async per second 2536.89 +- 17.32
multi client actor calls async per second 3695.97 +- 51.48
single client direct actor calls async per second 6909.92 +- 97.7
multi client direct actor calls async per second 9321.2 +- 186.79
multi client direct actor calls with arg async per second 4290.92 +- 372.21
```

### CODE REVIEW

File: src/ray/core_worker/core_worker.cc [hunk 1]
```diff
@@ -131,6 +132,9 @@ CoreWorker::CoreWorker(const WorkerType worker_type, const Language language,
       raylet_socket, WorkerID::FromBinary(worker_context_.GetWorkerID().Binary()),
       (worker_type_ == ray::WorkerType::WORKER), worker_context_.GetCurrentJobID(),
       language_, worker_server_.GetPort()));
+  if (direct_actor_task_receiver_ != nullptr) {
+    direct_actor_task_receiver_->Init(*raylet_client_);
```
Comment (stephanie-wang) [inline src/ray/core_worker/core_worker.cc]: Why not just pass this into the constructor?

File: src/ray/rpc/worker/direct_actor_client.h [hunk 2]
```diff
@@ -124,6 +129,10 @@ class DirectActorClient : public std::enable_shared_from_this<DirectActorClient>
 
   /// The max sequence number we have processed responses for.
   int64_t max_finished_seq_no_ GUARDED_BY(mutex_) = -1;
+
+  /// The task id we are currently sending requests for. When this changes,
+  /// the max finished seq no counter is reset.
+  std::string cur_caller_id_;
```
Comment (stephanie-wang) [inline src/ray/rpc/worker/direct_actor_client.h]: Was this a bug?

Also, shouldn't this be a TaskID?

File: src/ray/raylet/node_manager.cc [hunk 3]
```diff
@@ -1253,6 +1256,37 @@ void NodeManager::ProcessWaitRequestMessage(
...
+      // Add any missing objects to the list to subscribe to in the task
+      // dependency manager. These objects will be pulled from remote node
+      // managers and reconstructed if necessary.
+      required_object_ids.push_back(object_id);
+    }
+  }
+
+  ray::Status status = object_manager_.Wait(
+      object_ids, -1, object_ids.size(), false,
+      [this, client, tag](std::vector<ObjectID> found, std::vector<ObjectID> remaining) {
```
Comment (stephanie-wang) [inline src/ray/raylet/node_manager.cc]: I believe this will cause a memory leak if the worker dies and the object never appears.

File: src/ray/core_worker/transport/direct_actor_transport.h [hunk 4]
```diff
@@ -125,60 +126,139 @@ class CoreWorkerDirectActorTaskSubmitter {
...
+        has_pending_dependencies_(has_dependencies) {}
+
+  void Accept() { accept_callback_(); }
+  void Cancel() { reject_callback_(); }
+  bool CanExecute() { return !has_pending_dependencies_; }
+  void MarkDependenciesSatisfied() { has_pending_dependencies_ = false; }
+
+ private:
+  std::function<void()> accept_callback_;
+  std::function<void()> reject_callback_;
```
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_actor_transport.h]: const

File: src/ray/core_worker/transport/direct_actor_transport.h [hunk 5]
```diff
@@ -125,60 +126,139 @@ class CoreWorkerDirectActorTaskSubmitter {
...
+
+  void Wait(const std::vector<ObjectID> &dependencies,
+            std::function<void()> on_dependencies_available) override {
+    auto tag = next_request_id_++;
+    requests_[tag] = on_dependencies_available;
+    raylet_client_.WaitForDirectActorCallArgs(dependencies, tag);
+  }
+
+  /// Fulfills the callback stored by Wait().
+  void OnWaitComplete(int64_t tag) {
```
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_actor_transport.h]: This seems like it should also be part of the interface.

**Review by stephanie-wang — COMMENTED** [review]
What if we did not use request numbers for each `Wait` request and just sent back notifications for each object ID that becomes local? The request number logic is a bit complicated to follow right now, I think because you have to trace through multiple hashmaps and callbacks to figure out what's going on in the `SchedulingQueue`.

Also, is there a way we can reuse some logic from the `TaskDependencyManager` to figure out when a task is ready to run?

File: src/ray/core_worker/transport/direct_actor_transport.h [hunk 6]
```diff
@@ -227,6 +316,8 @@ class CoreWorkerDirectActorTaskReceiver : public rpc::DirectActorHandler {
   TaskHandler task_handler_;
   /// The IO event loop for running tasks on.
   boost::asio::io_service &task_main_io_service_;
+  /// Shared waiter for dependencies required by incoming tasks.
+  std::unique_ptr<DependencyWaiterImpl> waiter_;
```
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_actor_transport.h]: Does this need to be a unique_ptr?

File: src/ray/core_worker/transport/raylet_transport.h [hunk 7]
```diff
@@ -32,6 +33,12 @@ class CoreWorkerRayletTaskReceiver : public rpc::WorkerTaskHandler {
                         rpc::AssignTaskReply *reply,
                         rpc::SendReplyCallback send_reply_callback) override;
 
+  // TODO(ekl) this shouldn't be in the raylet task receiver
```
Comment (stephanie-wang) [inline src/ray/core_worker/transport/raylet_transport.h]: Why does this need to be here?

File: src/ray/core_worker/core_worker.cc (hunk 1 above)
Comment (ericl) [reply src/ray/core_worker/core_worker.cc]: Figured this out.

File: src/ray/core_worker/transport/direct_actor_transport.cc [hunk 8]
```diff
@@ -208,28 +208,38 @@ CoreWorkerDirectActorTaskReceiver::CoreWorkerDirectActorTaskReceiver(
...
-  }
   if (task_spec.IsActorTask() && !worker_context_.CurrentActorUseDirectCall()) {
     send_reply_callback(Status::Invalid("This actor doesn't accept direct calls."),
                         nullptr, nullptr);
     return;
   }
 
+  // TODO(ekl) resolving object dependencies is expensive and requires an IPC to
+  // the raylet, which is a central bottleneck. In the future, we should inline
+  // dependencies that are small and already known to be local to the client.
```
[the remark this answers is not in the document]
Comment (ericl) [reply src/ray/core_worker/transport/direct_actor_transport.cc]: I considered it, but this is quite a bit simpler and the performance hit isn't huge since raylet dedups internally. In the future we can inline small objects so no wait is required at all.

For local objects, I think with ownership you will have that data right? At least for owned objects.

File: src/ray/core_worker/transport/direct_actor_transport.h [hunk 9]
```diff
@@ -125,60 +126,139 @@ class CoreWorkerDirectActorTaskSubmitter {
...
+  InboundRequest(){};
+  InboundRequest(std::function<void()> accept_callback,
+                 std::function<void()> reject_callback, bool has_dependencies)
+      : accept_callback_(accept_callback),
+        reject_callback_(reject_callback),
+        has_pending_dependencies_(has_dependencies) {}
+
+  void Accept() { accept_callback_(); }
+  void Cancel() { reject_callback_(); }
+  bool CanExecute() { return !has_pending_dependencies_; }
```
[the remark this answers is not in the document]
Comment (ericl) [reply src/ray/core_worker/transport/direct_actor_transport.h]: Done.

File: src/ray/core_worker/transport/direct_actor_transport.h (hunk 4 above)
Comment (ericl) [reply src/ray/core_worker/transport/direct_actor_transport.h]: Seems marking this as const prevents copy construction...

File: src/ray/core_worker/transport/direct_actor_transport.h (hunk 5 above)
Comment (ericl) [reply src/ray/core_worker/transport/direct_actor_transport.h]: It's an implementation detail of this particular one, the mock doesn't require it.

File: src/ray/core_worker/transport/direct_actor_transport.h (hunk 6 above)
Comment (ericl) [reply src/ray/core_worker/transport/direct_actor_transport.h]: Yeah, you can't declare a polymorphic class inline.

File: src/ray/core_worker/transport/raylet_transport.h (hunk 7 above)
Comment (ericl) [reply src/ray/core_worker/transport/raylet_transport.h]: I moved it. Since the worker and direct actor services are separate, this means we now need two rpc clients to each worker, which is a bit weird but probably ok since clients are lightweight.

File: src/ray/raylet/node_manager.cc [hunk 10]
```diff
@@ -1253,6 +1256,37 @@ void NodeManager::ProcessWaitRequestMessage(
...
+  for (auto const &object_id : object_ids) {
+    if (!task_dependency_manager_.CheckObjectLocal(object_id)) {
+      // Add any missing objects to the list to subscribe to in the task
+      // dependency manager. These objects will be pulled from remote node
+      // managers and reconstructed if necessary.
+      required_object_ids.push_back(object_id);
+    }
+  }
+
+  ray::Status status = object_manager_.Wait(
```
[the remark this answers is not in the document]
Comment (ericl) [reply src/ray/raylet/node_manager.cc]: I was assuming Wait() would still return if there is a failure, and then the Get() would raise an exception later. Is this not the case?

File: src/ray/raylet/node_manager.cc (hunk 3 above)
Comment (ericl) [reply src/ray/raylet/node_manager.cc]: The wait() will always complete right?

File: src/ray/rpc/worker/direct_actor_client.h (hunk 2 above)
Comment (ericl) [reply src/ray/rpc/worker/direct_actor_client.h]: Yep. It's encoded as binary -- no need to decode.

**Review by ericl — COMMENTED** [review]
Thanks @stephanie-wang , updated.

File: src/ray/core_worker/core_worker.cc (hunk 1 above)
Comment (ericl) [reply src/ray/core_worker/core_worker.cc]: Oh actually, we still need this. The reason is the raylet client cannot be constructed prior to the receivers.

File: src/ray/raylet/node_manager.cc (hunk 10 above)
[the remark this answers is not in the document]
Comment (ericl) [reply src/ray/raylet/node_manager.cc]: I just double checked and you do get the right exception on failure. Added a test for this.

File: src/ray/protobuf/direct_actor.proto [hunk 11]
```diff
@@ -33,8 +33,20 @@ message PushTaskReply {
   repeated ReturnObject return_objects = 1;
 }
 
+message DirectActorCallArgWaitCompleteRequest {
```
Comment (pcmoritz) [inline src/ray/protobuf/direct_actor.proto]: This would also be very useful for @suquark s ray `async_api`. Using it could allow us to get rid of the last remaining python plasma client in ray :)

File: src/ray/core_worker/core_worker.cc (hunk 1 above)
Comment (edoakes) [reply src/ray/core_worker/core_worker.cc]: Why can't raylet client be constructed before the receivers? Leave a comment pls.

File: src/ray/core_worker/transport/direct_actor_transport.cc [hunk 12]
```diff
@@ -84,7 +81,7 @@ void CoreWorkerDirectActorTaskSubmitter::HandleActorUpdate(
     // Check if this actor is the one that we're interested, if we already have
```
Comment (edoakes) [inline src/ray/core_worker/transport/direct_actor_transport.cc]: Not part of this PR explicitly but this comment is very confusing. Can you clarify it please?

Comment (edoakes) [reply src/ray/core_worker/transport/direct_actor_transport.cc]: It is unintuitive to me that we would create a new connection if we have an existing one.

File: src/ray/core_worker/transport/direct_actor_transport.cc [hunk 13]
```diff
@@ -147,6 +144,9 @@ void CoreWorkerDirectActorTaskSubmitter::PushTask(
           waiting_reply_tasks_[actor_id].erase(task_id);
         }
         if (!status.ok()) {
+          // Note that this might be the __ray_terminate__ task, so we don't log
```
Comment (edoakes) [inline src/ray/core_worker/transport/direct_actor_transport.cc]: Shouldn't __ray_terminate__ just return `Status::OK`?

File: src/ray/core_worker/transport/direct_actor_transport.h [hunk 14]
```diff
@@ -125,60 +126,139 @@ class CoreWorkerDirectActorTaskSubmitter {
   friend class CoreWorkerTest;
 };
 
+/// Object dependency and RPC state of an inbound request.
+class InboundRequest {
+ public:
+  InboundRequest(){};
```
Comment (edoakes) [inline src/ray/core_worker/transport/direct_actor_transport.h]: Is this used?

File: src/ray/core_worker/transport/direct_actor_transport.h [hunk 15]
```diff
@@ -208,16 +291,27 @@ class CoreWorkerDirectActorTaskReceiver : public rpc::DirectActorHandler {
...
+  void Init(RayletClient &client);
+
   /// Handle a `PushTask` request.
-  /// The implementation can handle this request asynchronously. When hanling is done, the
-  /// `done_callback` should be called.
   ///
   /// \param[in] request The request message.
   /// \param[out] reply The reply message.
-  /// \param[in] done_callback The callback to be called when the request is done.
+  /// \param[in] send_replay_callback The callback to be called when the request is done.
```
Comment (edoakes) [inline src/ray/core_worker/transport/direct_actor_transport.h]:
```suggestion
  /// \param[in] send_reply_callback The callback to be called when the request is done.
```

File: src/ray/core_worker/transport/direct_actor_transport.h [hunk 16]
```diff
@@ -208,16 +291,27 @@ class CoreWorkerDirectActorTaskReceiver : public rpc::DirectActorHandler {
...
   void HandlePushTask(const rpc::PushTaskRequest &request, rpc::PushTaskReply *reply,
                       rpc::SendReplyCallback send_reply_callback) override;
 
+  /// Handle a `DirectActorCallArgWaitComplete` request.
+  ///
+  /// \param[in] request The request message.
+  /// \param[out] reply The reply message.
+  /// \param[in] send_replay_callback The callback to be called when the request is done.
```
Comment (edoakes) [inline src/ray/core_worker/transport/direct_actor_transport.h]:
```suggestion
  /// \param[in] send_reply_callback The callback to be called when the request is done.
```

File: src/ray/raylet/node_manager.cc [hunk 17]
```diff
@@ -1253,6 +1256,37 @@ void NodeManager::ProcessWaitRequestMessage(
   RAY_CHECK_OK(status);
 }
 
+void NodeManager::ProcessWaitForDirectActorCallArgsRequestMessage(
```
Comment (edoakes) [inline src/ray/raylet/node_manager.cc]: Can you rename this to something more general like `AsyncWait`? Not really tied to DAC at all.

File: src/ray/raylet/node_manager.cc (hunk 3 above)
Comment (edoakes) [reply src/ray/raylet/node_manager.cc]: Should confirm this before merging.

**Review by edoakes — APPROVED** [review]
LGTM

File: src/ray/core_worker/transport/direct_actor_transport.cc (hunk 12 above)
Comment (ericl) [reply src/ray/core_worker/transport/direct_actor_transport.cc]: Pending requests means we don't have any existing connection.

File: src/ray/core_worker/transport/direct_actor_transport.cc (hunk 13 above)
Comment (ericl) [reply src/ray/core_worker/transport/direct_actor_transport.cc]: I have a fix for this in an upcoming PR...

File: src/ray/core_worker/transport/direct_actor_transport.h (hunk 14 above)
Comment (ericl) [reply src/ray/core_worker/transport/direct_actor_transport.h]: Yes, for default initialization.

**Review by ericl — COMMENTED** [review]
updated
